API Separate searchupdate / commit into separate queued-jobs

tractorcow opened the pull request:
Oh my good, look at this pull request. It's so amazing!

What have I done! On a coffee-less afternoon nonetheless?

- Indexes can finally hide other indexes. This is useful when you don't want classes to be indexed, but they could be in a third party module you don't want to edit. (see docs below) This in itself is a great performance increase, but no longer do you need to have solr indexing work done on indexes you don't want to maintain.

- SolrIndex::commit has been split into a separate queued job which is added to the queue once an update job has been completed. This is useful if you want to run multiple update jobs in a sequence, but don't want to run into the dreaded `exceeded limit of maxWarmingSearchers=X` error.

It also nicely bunches several smaller updates into a single commit, so hopefully less work is done overall.

Also if for whatever reason a commit is made in a single job, and updates occur later (again, requiring further updates) then the job will be re-added to the queue with a nice cooldown (defaults to 10 minutes) to prevent solr erroring, and giving time for the rest of the job queue to complete.

chillu commented on code/search/FullTextSearch.php:
```diff
@@ -18,16 +18,41 @@ class FullTextSearch {
+				$candidates = array();
 				foreach ($classes as $class) {
+					// Check if this index is disabled
+					$hides = $class::config()->hide_ancestor;
```
I would prefer to make this an explicit registration system, less magic and more versatile. The YAML configuration doesn't work because we can't overwrite lists, but we could use `Config::update()` for that?

```yml
FullTextSearch:
  indexes:
    - MySolrIndex
```

tractorcow replied:
I wanted to go for something that, if ignored, doesn't have any effect, and is fully backwards compatible.

You can do the same in yaml though.

```
MyNewIndex:
  hide_ancestor: MyOldIndex
```

Ok, maybe I can split it into a new PR.

tractorcow replied:
Ok I'm going to add both; You can use an 'indexes' config to do a pre-filter, where items can be subsequently hidden via hide_ancestor.

phptek commented:
@tractorcow If this works I will have your babies.

patbolo commented on tests/BatchedProcessorTest.php:
```diff
@@ -100,7 +138,60 @@ public function testBatching() {
+		$this->assertFalse($first->jobFinished());
+		$first->process();
+		$allMessages = $first->getMessages();
+		$this->assertTrue($index->getIsCommitted());
```
Shouldn't the commit be performed after $second has been processed?

tractorcow replied:
If the first commit has run, it should be committed then, not after the `$second->process()`.

When the second one is run it should see that there is in fact no more committing left to do, hence the "discarding this job" message.

tractorcow commented on tests/BatchedProcessorTest.php:
```diff
@@ -100,7 +138,60 @@ public function testBatching() {
+		$this->assertFalse($second->jobFinished());
+		$second->process();
+		$allMessages = $second->getMessages();
+		$this->assertFalse($index->getIsCommitted());
```
@patbolo I only temporarily reset the $index to set comitted to false to prove that it's not re-comitted by `$second`. In practice it would stay comitted, although just for the sake of this test we toggle it back to false.

normann commented:
@phptek lust comment

tractorcow commented:
So these have been changed:
- Implemented @chillu's 'FulltextSearch.indexes' config option to globally filter indexes
- Reduced cooldown on the commit task to 5 minutes
- rebased :D

assertchris commented on code/search/SearchIndex.php:
```diff
@@ -534,10 +542,12 @@ function commit() { }
 	public $deleted = array();
+	public $committed = false;
 
 	function reset() {
```
Maybe add some visibility keywords... #boyscout

tractorcow replied:
Done.

assertchris commented on code/search/processors/SearchUpdateCommitJobProcessor.php:
```diff
@@ -0,0 +1,247 @@
+		// NOOP
+	}
+
+	public function getJobData() {
```
The name of this method is very misleading (especially given the following method's name). Perhaps `getNewJobObject`?

tractorcow replied:
It's an interface method on QueuedJob, so, I can't change it. =(

tractorcow commented:
@chillu is just running a test of this code on 15k documents, so let's see how that turns out. ;D

tractorcow commented:
I've added a fix to prevent lots of unnecessary duplicate commit jobs being added to the queue.

chillu commented:
Had a looooong discussion and debugging session with @tractorcow on a large production data set around this pull request. 

The logic could be improved in terms of queueing - it currently starts creating "update" jobs synchronously until the first job actually gets picked up for execution. Since the update job completion triggers the first (and only) "commit" job to be queued, it purely depends on the server's execution speed on how many "update" jobs are in the queue before the first "commit" job. Ideally this would be a fixed interval, depending on the performance of the search warmers (which might be a separate box with altogether separate performance characteristics). But that's for another pull request.

As usual, *legend* Mr. Mooyman!
